src/python/pants/core/goals/tailor.py

```diff
@@ -80,6 +80,19 @@ def default_sources_for_target_type(tgt_type: type[Target]) -> tuple[str, ...]:
     return tuple()
 
 
+@memoized
+def has_source_or_sources_field(tgt_type: type[Target]) -> bool:
+    """Tell whether a given target type has a `source` or `sources` field.
+
+    This may be useful when determining whether it's possible to tailor a target with the passed
+    source(s) field value if the target doesn't have such a field in the first place.
+    """
+    for field in tgt_type.core_fields:
+        if issubclass(field, (OptionalSingleSourceField, MultipleSourcesField)):
```

Comment:
These two classes come from https://github.com/pantsbuild/pants/pull/18174/files#diff-95512e3b746327a940d69c10e6d376f746d57a19b97c7f7cb9abc87ee68e8863R76-R79, so I thought they are safe to use.

Reply:
I think it's better to use the common base class for these, so no need to loop here, just:
```python
return tgt_type.has_field(SourcesField)
```

And then I wonder if it really is worth memoizing here..?
Feels like if this is costly and needs optimizing, it would make more sense to memoize closer to `Target.has_field()`.

Reply:
I saw this is how it's done in

```
def default_sources_for_target_type(tgt_type: type[Target]) -> tuple[str, ...]:
    for field in tgt_type.core_fields:
        if issubclass(field, OptionalSingleSourceField):
            return (field.default,) if field.default else tuple()
        if issubclass(field, MultipleSourcesField):
            return field.default or tuple()
    return tuple()
```

so I figured that is the right way? Pardon my limited understanding of the API, but I don't think `tgt_type.has_field(SourcesField)` is possible to do on a `type[Target]`?

Reply:
it is--and also the example you pulled does different things depending on if it is a single vs multi source field where as you do not ;)
https://github.com/pantsbuild/pants/blob/2983f166e18b1c1a480b85bf722d16d514b41b17/src/python/pants/engine/target.py#L1929-L1932

Reply:
Hm it doesn't seem to be possible in my case:

```
  File "/home/alexey.tereshenkov/code/pants/src/python/pants/core/goals/tailor.py", line 89, in has_source_or_sources_field
    return tgt_type.has_field(SourcesField)
TypeError: has_field() missing 1 required positional argument: 'field'
```

with 

```diff
diff --git a/src/python/pants/core/goals/tailor.py b/src/python/pants/core/goals/tailor.py
index d5a1c1309..8180fde2d 100644
--- a/src/python/pants/core/goals/tailor.py
+++ b/src/python/pants/core/goals/tailor.py
@@ -86,10 +86,7 @@ def has_source_or_sources_field(tgt_type: type[Target]) -> bool:
     This may be useful when determining whether it's possible to tailor a target with the passed
     source(s) field value if the target doesn't have such a field in the first place.
     """
-    for field in tgt_type.core_fields:
-        if issubclass(field, (OptionalSingleSourceField, MultipleSourcesField)):
-            return True
-    return False
+    return tgt_type.has_field(SourcesField)
 
 
 @dataclass(order=True, frozen=True)
```

Reply:
> the example you pulled does different things depending on if it is a single vs multi source field where as you do not ;)

Sorry should have clarified - I followed this pattern since I got the error on the first attempt to do `tgt_type.has_field(SourcesField)` I'd normally do on an instance of a `Target`

Reply:
Huh... ok that to me was unexpected. I'll 👀 into it.. (no need to block this PR on it, tho)

Reply:
Sorry, my bad, for classes there is: 
https://github.com/pantsbuild/pants/blob/817db4395a45b4de18b37a513e71d8393c7c34a1/src/python/pants/engine/target.py#L672-L675

Reply:
Ah right. `return tgt_type.class_has_field(SourcesField, UnionMembership({}))` seems to be working. I don't fully understand the purpose of bringing the union of rules here, but that's fine. Would this be a semantically correct call to make?

Reply:
yes, except you'd need to provide the actual unionmembership instance rather than creating your own. (you get it as a type to your rule, i.e. `async def my_rule(..., union_membership: UnionMembership, ...)`).

The reason it's needed is to make any plugin fields available.

Reply:
maybe leave a note saying that this will not work for targets with a plugin source field (which is rare/not an issue in practice I guess)